### PR TITLE
HPCC 26271 update Tensorflow for gpu support

### PR DESCRIPTION
**Pull request by xwang2713**
Also update README.md for instruction about how to build and test the Docker image.

Signed off by Xiaoming Wang@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [x] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Build and test in Azure AKS

### CODE REVIEW

**Comment by xwang2713** [comment]
@g-pan please review the README.md. I re-write it.

**Comment by xwang2713** [comment]
@bobf2000 there is a test Docker image: hpccsystemslegacy/platform-gnn-gpu:8.8.0-rc8
To test GPU and CUDA driver:
Python3:
    import tensorflow as tf
    print(tf.test.is_gpu_available())
    print(tf.test.is_gpu_available(cuda_only=True))

Reference JIRA HPCC-26271 for more detail

**Comment by xwang2713** [comment]
I need update CUDA toolkit in the Docker image since CUDA toolkit 11.3 isn't compatible with GPU Driver on the NC6S-V3 card which has "NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4"
https://docs.nvidia.com/deploy/cuda-compatibility/index.html

File: dockerfiles/platform-gnn-gpu/README.md [hunk 1]
```diff
@@ -1,45 +1,100 @@
 # HPCC Systems Machine Learning GNN GPU Docker Images
 
-## Tensorflow 2 with GPU support
+## Tensorflow with GPU support
 
-The new Tensorflow 2 and Nvidia Cuda libraries are not always compatible. It will be ideal if we can use Docker image by published by Tensorflow (Docker Hub tensorflow/tensorflow <version>-gpu) but the base image is Ubuntu 18.04 and we use Ubuntu 20.04. So instead we create the images by addiing Tensorflow 2 and Nvida Cuda, etc libraries on the top of hpccystems/platform-core image.
+### Support Tensorflow on a GPU with Docker
+Docker is the easiest way to run TensorFlow on a GPU since the host machine only requires the NVIDIA® driver (the NVIDIA® CUDA® Toolkit is not required).
+When you create a AKS cluster with GPU support VM size NVIDIA driver is installed on the nodes.
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:7]: When you create an AKS cluster with GPU support a VM sized NVIDIA driver is installed on the nodes.

File: dockerfiles/platform-gnn-gpu/README.md [hunk 2]
```diff
@@ -1,45 +1,100 @@
...
+Run "nvidia-smi" will print the NVIDIA driver version. Our HPCC GNN GPU Docker image, platform-gnn-gpu, will have NVIDIA CUDA Toolkit.
+Run "nvcc --version" will print the version. Beware the compatibility issue. For example CUDA Toolkit 11.3 and earlier will not be compatible with NVIDIA Driver 11.4+.  Reference https://docs.nvidia.com/deploy/cuda-compatibility/index.html for more information.
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:9]: Run "nvcc --version" to display the version.

File: dockerfiles/platform-gnn-gpu/README.md [hunk 3]
```diff
@@ -1,45 +1,100 @@
...
+Current NCS V3 node has  NVIDIA Driver 11.4.
 
-When preparing the Dockerfile, specially for Tensorflow 2 with GPU support it is important to reference
-Tensorflow and Nvidia Cuda Dockerfiles and Docker images to pick compatible libabries.
+### GPU support in our Azure subscription
 
-### Docker image and Dockerfile Reference
+Our Azure subscription does include VM Size of GPU support. These VM Sizes normally are much more expensive than normal VM Sizes, probaly at least 10 times.  Each these VM Sizes with GPU support have different hardware and may require different driver version such as NVidia CUDA driver. Also the Tensorflow 2 and Nvidia Cuda libraries are not always compatible.
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:14]: Our Azure subscription does include VM Size of GPU support. These VM Sizes normally are much more expensive than normal VM Sizes, probably at least 10 times.  Each of these VM Sizes with GPU support have different hardware and may require different driver versions, such as NVidia CUDA driver. Also the Tensorflow 2 and Nvidia Cuda libraries are not always compatible.

File: dockerfiles/platform-gnn-gpu/README.md [hunk 4]
```diff
@@ -1,45 +1,100 @@
...
 
-#### Tensforflow
+Generally  some ND, NCv2 or NCv3 family VM Sizes should be supported in our Azure subscription
+For example following are supported: standard_nc6s_v3,standard_nc8as_t4_v3,standard_nv12s_v3,standard_nv24s_v3,standard_nv48s_v3
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:17]: For example, the following are supported:

File: dockerfiles/platform-gnn-gpu/README.md [hunk 5]
```diff
@@ -1,45 +1,100 @@
...
 
-- Dockerfile:
-  https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/dockerfiles/dockerfiles/gpu.Dockerfile
-  Above Dockeerfile is based https://hub.docker.com/layers/nvidia/cuda/. There are runime, base, etc.
-  For now we just reference https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/ and replace the package version for Ubuntu 20.04
+We current build our Docker Image with NVidia CUDA 11.4 and latest Tensorflow packages. User may need to create their own Docker Images with a different CUDA driver version and Tensorflow version.
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:19]: We currently build our Docker Image with NVidia CUDA 11.4 and the latest Tensorflow packages. Users may need to create their own Docker Images with a different CUDA driver version and Tensorflow version.

File: dockerfiles/platform-gnn-gpu/README.md [hunk 6]
```diff
@@ -1,45 +1,100 @@
...
 
-- The generated tensorflow/tensorflow:2.2.0-gpu
-  https://hub.docker.com/layers/tensorflow/tensorflow/2.2.0-gpu/images/sha256-3f8f06cdfbc09c54568f191bbc54419b348ecc08dc5e031a53c22c6bba0a252e?context=explore
+When preparing the Dockerfile, specially for Tensorflow 2 with GPU support it is important to reference
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:21]: When preparing the Dockerfile, especially for Tensorflow 2 with GPU support it is important to reference

File: dockerfiles/platform-gnn-gpu/README.md [hunk 7]
```diff
@@ -1,45 +1,100 @@
...
+Tensorflow and Nvidia Cuda Dockerfiles and Docker images to pick compatible libabries.
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:22]: Tensorflow and Nvidia Cuda Dockerfiles and Docker images to pick compatible libraries.

File: dockerfiles/platform-gnn-gpu/README.md [hunk 8]
```diff
@@ -1,45 +1,100 @@
...
+- NVIDIA GPU Cloud (NGC) https://www.nvidia.com/en-us/gpu-cloud/
 
-  https://hub.docker.com/layers/nvidia/cuda/11.3.1-base/images/sha256-351d731f46159c1afdb003121e96b512a77c410c2b43ee6ea74cf1413bb858ab?context=explore
 
-- NGC-DL-CONTAINER-LICENSE
-  https://gitlab.com/nvidia/container-images/cuda/-/blob/master/LICENSE under BSD 3-Clause "New" or "Revised" License
+#### Dockerfile Consideration
+Tensorflow only provides CUDA 11.2 Dockerfile for GPU support: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/dockerfiles/dockerfiles/gpu.Dockerfile
 
+When preparing our Dockerfile (CUDA 11.4) we reference above gpu.Dockerfile with following modification
```
Comment (g-pan) [inline dockerfiles/platform-gnn-gpu/README.md:35]: When preparing our Dockerfile (CUDA 11.4) we reference the above gpu.Dockerfile with following modifications

**Review by g-pan — COMMENTED** [review]
A few minor changes noted in line.

File: dockerfiles/platform-gnn-gpu/README.md (hunk 1 above)
Comment (xwang2713) [reply dockerfiles/platform-gnn-gpu/README.md:7]: Fixed

File: dockerfiles/platform-gnn-gpu/README.md (hunk 2 above)
Comment (xwang2713) [reply dockerfiles/platform-gnn-gpu/README.md:9]: Fixed

File: dockerfiles/platform-gnn-gpu/README.md (hunk 3 above)
Comment (xwang2713) [reply dockerfiles/platform-gnn-gpu/README.md:14]: Fixed

File: dockerfiles/platform-gnn-gpu/README.md (hunk 4 above)
Comment (xwang2713) [reply dockerfiles/platform-gnn-gpu/README.md:17]: Fixed

File: dockerfiles/platform-gnn-gpu/README.md (hunk 7 above)
Comment (xwang2713) [reply dockerfiles/platform-gnn-gpu/README.md:22]: Fixed

**Review by g-pan — APPROVED** [review]

**Comment by richardkchapman** [comment]
@xwang2713 Please squash

**Comment by xwang2713** [comment]
@richardkchapman squash done

**Pull request merged**